onnxruntime/core/graph/graph.cc

```diff
@@ -989,6 +989,30 @@ Graph::Graph(const Model& owning_model,
                      }),
       graph_mutable_nodes->end());
 
+  // For now we convert sparse_intializer to dense tensors
+  // since there are currently no supported ops that consume sparse
+  // initializers directly.
+  // Unlike Constant nodes we are not removing sparse initializers from the graph
+  // unless they are not used by any operation. This is done so we can still save space
+  // when storing an optimized model in ort format.
+  if (graph_proto_->sparse_initializer_size() > 0) {
+    for (const auto& sparse_tensor : graph_proto_->sparse_initializer()) {
+      ORT_ENFORCE(utils::HasName(sparse_tensor), "Sparse initializer must have a name. This model is invalid");
+      const gsl::not_null<TensorProto*> tensor{graph_proto_->add_initializer()};
+      auto status = utils::SparseTensorProtoToDenseTensorProto(sparse_tensor, *tensor);
+      ORT_ENFORCE(status.IsOK(), status.ToString());
+      auto p = sparse_tensor_names_.emplace(tensor->name());
+      ORT_ENFORCE(p.second, "Duplicate sparse_tensor_initializer: '", tensor->name(), "' Model is invalid.");
```

Comment:
Could you check duplicate before you convert the sparse tensor to dense?

Reply:
No, I can not. sparse_tensor_names_ takes a reference to name so the name should continue to exist after sparse initializer is gone and that means conversion.

Reply:
Should be exceedingly rare for there to be a duplicate as well. Wouldn't be a valid onnx model as you can't have two values with the same name.